File: NeetCode150/Arrays&Hashing/group_anagram.py
class Solution(object):
    def groupAnagrams(self, strs):
        res = []
        map = {}

        for str in strs:
            sorted_str = ''.join(sorted(str.lower()))

            if sorted_str in map:
                map[sorted_str].append(str)
            else:
                map[sorted_str] = [str]
        
        for val in map.values():
            res.append(val)
        
        return res

File: LeetCode2024/Array/GroupAnagram.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for (String str : strs) {
            char[] c = str.toCharArray();
            Arrays.sort(c);
            String sorted = new String(c);
            if (! map.containsKey(sorted)) {
                List<String> list = new ArrayList<>();
                list.add(str); // original str (non-sorted)
                map.put(sorted, list);
            } else {
                List<String> list = map.get(sorted);
                list.add(str);
                map.put(sorted, list);
            }
        }

        List<List<String>> res = new ArrayList<>();

        map.forEach((k, v) -> res.add(new ArrayList<>(v)));

        return res;
    }
}
